api/v1beta1/vmscrapeconfig_types.go

```diff
@@ -0,0 +1,557 @@
+	// Configure whether HTTP requests follow HTTP 3xx redirects.
+	// +optional
+	FollowRedirects *bool `json:"followRedirects,omitempty"`
+	// Optional namespace discovery. If omitted, discover targets across all namespaces.
+	// +optional
+	Namespaces *NamespaceDiscovery `json:"namespaces,omitempty"`
+	// AttachMetadata configures metadata attaching from service discovery
+	// +optional
+	AttachMetadata AttachMetadata `json:"attach_metadata,omitempty"`
+	// Selector to select objects.
+	// +optional
+	// +listType=map
+	// +listMapKey=role
+	Selectors []K8SSelectorConfig `json:"selectors,omitempty"`
+}
+
+// K8SSelectorConfig is Kubernetes Selector Config
+type K8SSelectorConfig struct {
+	// +kubebuilder:validation:Required
+	Role  string `json:"role"`
+	Label string `json:"label,omitempty"`
+	Field string `json:"field,omitempty"`
+}
+
+// NamespaceDiscovery is the configuration for discovering
+// Kubernetes namespaces.
+type NamespaceDiscovery struct {
+	// Includes the namespace in which the pod exists to the list of watched namespaces.
+	// +optional
+	IncludeOwnNamespace *bool `json:"ownNamespace,omitempty"`
+	// List of namespaces where to watch for resources.
+	// If empty and `ownNamespace` isn't true, watch for resources in all namespaces.
+	// +optional
+	Names []string `json:"names,omitempty"`
+}
+
+// ConsulSDConfig defines a Consul service discovery configuration
+// See https://docs.victoriametrics.com/sd_configs/#consul_sd_configs
+// +k8s:openapi-gen=true
+type ConsulSDConfig struct {
+	// A valid string consisting of a hostname or IP followed by an optional port number.
+	// +kubebuilder:validation:MinLength=1
+	// +required
+	Server string `json:"server"`
+	// Consul ACL TokenRef, if not provided it will use the ACL from the local Consul Agent.
+	// +optional
+	TokenRef *v1.SecretKeySelector `json:"tokenRef,omitempty"`
+	// Consul Datacenter name, if not provided it will use the local Consul Agent Datacenter.
+	// +optional
+	Datacenter *string `json:"datacenter,omitempty"`
+	// Namespaces are only supported in Consul Enterprise.
+	// +optional
+	Namespace *string `json:"namespace,omitempty"`
+	// Admin Partitions are only supported in Consul Enterprise.
+	// +optional
+	Partition *string `json:"partition,omitempty"`
+	// HTTP Scheme default "http"
+	// +kubebuilder:validation:Enum=HTTP;HTTPS
+	// +optional
+	Scheme *string `json:"scheme,omitempty"`
+	// A list of services for which targets are retrieved. If omitted, all services are scraped.
+	// +listType:=atomic
+	// +optional
+	Services []string `json:"services,omitempty"`
+	// An optional list of tags used to filter nodes for a given service. Services must contain all tags in the list.
+	//+listType:=atomic
+	// +optional
+	Tags []string `json:"tags,omitempty"`
+	// The string by which Consul tags are joined into the tag label.
+	// If unset, use its default value.
+	// +optional
+	TagSeparator *string `json:"tagSeparator,omitempty"`
```

Comment:
Should we have `yaml` related tags for the fields? 

For instance, `TagSeparator` has  definition: `yaml:"tag_separator,omitempty"`  at `vmagent` config.

Reply:
Didn't see the necessity for it, since we're not using yaml library for vmscrapeconfig.
And we have mixed tag usage in various types now, with yaml tag or without it, maybe we should sync them all.